gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/CopySourceWithMultipleSourceLocation.java

```diff
@@ -0,0 +1,276 @@
+
+      List<URI> replicationDataset = new ArrayList<URI>(importedBy);
+      Comparator<URI> pathLengthComparator = new Comparator<URI>() {
+        public int compare(URI c1, URI c2) {
+          return c1.getPath().length() - c2.getPath().length();
+        }
+      };
+
+      // sort the URI based on the path length to make sure the parent path appear before children
+      Collections.sort(replicationDataset, pathLengthComparator);
+
+      Set<URI> leafDataset = new HashSet<URI>();
+
+      for (URI u : replicationDataset) {
+        // due to the sort, at most one element need to be removed 
+        URI needToRemove = null;
+
+        // valid dataset URI
+        if (u.getPath().startsWith(replicationRootPath)) {
+          for (URI leaf : leafDataset) {
+            if (PathUtils.isAncestor(new Path(leaf.getPath()), new Path(u.getPath()))) {
+              needToRemove = leaf;
+              break;
+            }
+          }
+
+          if (needToRemove != null) {
+            leafDataset.remove(needToRemove);
+          }
+          leafDataset.add(u);
+        }
+      }
+
+      List<Config> result = new ArrayList<Config>();
+      for (URI u : leafDataset) {
+        result.add(configClient.getConfig(u));
+      }
+
+      return result;
+    } catch (VersionDoesNotExistException | ConfigStoreFactoryDoesNotExistsException | ConfigStoreCreationException
+        | URISyntaxException e) {
+      log.error("Caught error " + e.getMessage());
+      throw new RuntimeException(e);
+    }
+  }
+
+  private boolean generateWorkunitsPerReplicationDataset(SourceState originState,
+      final ConcurrentBoundedWorkUnitList workUnitList, final long minWorkUnitWeight, ReplicationConfiguration rc)
+      throws IOException {
+    SourceState stateCp = new SourceState();
+    stateCp.addAll(originState);
+    ReplicationMetaData md = rc.getMetaData();
+    log.info("metadata : " + md);
+
+    SourceEndPoint source = rc.getSource();
+    log.info("source : " + source);
+
+    List<ReplicaEndPoint> replicas = rc.getReplicas();
+    for (ReplicaEndPoint r : replicas) {
+      log.info("replica: " + r);
+    }
+
+    DataFlowTopology topology = rc.getTopology();
+    for (DataFlowTopology.CopyRoute route : topology.getRoutes()) {
+      log.info("route: " + route);
+    }
+
+    DataFlowTopology.CopyRoute route = this.getCopyRoute(topology);
+    log.info("route is " + route);
+
+    // TODO, for now only pick the first copyFrom
+    if (source.getReplicationLocation().getType() == ReplicationLocationType.HADOOPFS) {
+
+      ReplicationLocation copyTo = route.getCopyTo().getReplicationLocation();
+      ReplicationLocation copyFrom = route.getCopyFrom().get(0).getReplicationLocation();
+      stateCp.setProp(ConfigurationKeys.SOURCE_FILEBASED_FS_URI, copyFrom.getFsURI());
+      stateCp.setProp(ConfigurableGlobDatasetFinder.DATASET_FINDER_PATTERN_KEY, copyFrom.getRootPath());
+      log.info("source dir is " + copyFrom.getRootPath());
```

Comment:
Logging should provide enough context what it relates to. If you have multiple datasets being processed in parallel, it will be impossible to determine what this logging relates to.